File: leetcode_other/Greedy/gas_station.py
"""
https://www.youtube.com/watch?v=lJwbPZGo05A&ab_channel=NeetCode
https://leetcode.com/problems/gas-station/discuss/3011271/Python-3-oror-2-6-lines-w-explanation-and-example-oror-TM%3A-99-98
https://leetcode.com/problems/gas-station/discuss/1706142/JavaC%2B%2BPython-An-explanation-that-ever-EXISTS-till-now!!!!

Time: O(N)
Space: O(1)

Key Concepts (Greedy Approach):
1) It's only possible to complete the circuit if the total amount of gas on the circuit is sufficient
enough to drive the circuit (i.e sum(gas) >= sum(cost))
2) The starting position can be determined by starting at some station A, and noting whether
a station B on the circuit is unreachable due to lack of gas. If all are reachable,
then A is our answer. If not, our answer is not A, nor any station between between A and B
    -Say if A can't make it, A+1 ... A+2 ... to B cannot make it either, because for A+1, A+2...,
    they all had to go through A first, so if A can't make it, these can't either

3) Reset the tank to zero and repeat on the remainder of gas 
4) The last station that is unreachable in this process (i.e station Z) is our answer

Example:
Gas = [1,2,3,4,5]
Cost= [3,4,5,1,2]

sum(gas) = 15
sum(cost) = 15

There must be at least one answer because 15 >= 15

total = 0
res = 0

1st iteration
total += gas[0] - cost[0] = 1 - 3 = -2
total < 0

reset total = 0, res = 0 + 1

total = 0
res = 1

2nd iteration
total += gas[1] - cost[1] = 2 - 4 = -2
total < 0

reset total = 0, res = 1 + 1 = 2

total = 0
res = 2

3rd iteration
total += gas[2] - cost[2] = 3 - 5 = -2
total < 0

reset total, res = 2 + 1 = 3

4th iteration
total += gas[3] - cost[3] = 3
total > 0,

total = 3

5th iteration

total += gas[4] - cost[4] = 3 += 3 = 6

total = 6

Therefore, res is still equal to 3 which is the answer

Another example:
Gas = [1,2,3,4,5,4,1,1,1]
Cost= [3,4,5,1,2,1,2,10,8]

if running through this similar example above,
the total becomes non zero at i = 3, however,
the total dips back down below zero at i = 7, where the cost was 10

This is why we can't just return the 1st instance of the total
becoming non-zero, we need to run through the rest of the array, 
because at i = 8, the total becomes non-zero again,
making i = 8 our answer

"""
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        if sum(gas) < sum(cost):
            return -1
        total = 0
        res = 0
        for i in range(len(gas)):
            total += gas[i] - cost[i]
            if total < 0:
                total = 0
                res = i + 1
        return res
        

File: node_js_ds_and_algos_implementations/greedy/gas_station.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
    /*
    https://leetcode.com/problems/gas-station/discuss/805989/brute-force-one-pass-step-by-step-explanationwith-proof-graph-and-code-beginners-friendly
    Greedy Solution
    Similar to Kadane's Algorithm:

    1) 1st check is realizing that if the total gas < total cost, there is no valid answer.
    The reason is because if total gas < total cost, you will eventually get to a point
    where you don't have enough gas to leave one of the stations.

    2) 2nd part is similar to Kadane's. Keep track of the total amount of gas that would remain
    after getting to a station and attempting to leave (total += gas[i] - cost[i])

        if total < 0, that means that we know that any starting point up to this point would not be valid,
        as it'd run out of gas trying to leave station i. Therefore, we can reset the total gas to 0,
        and then set i + 1 to be the new starting point, since we know i won't work.

    Example:
    gas = [5,1,2,3,4]
    cost = [4,4,1,5,1]

    Time: O(N)
    Space: O(1)
    */
    let totalGas = gas.reduce((acc, x) => acc + x, 0)
    let totalCost = cost.reduce((acc, x) => acc + x, 0)
    if (totalGas < totalCost){
        return -1
    }
    totalGas = 0
    start = 0
    for (let i = 0; i < gas.length; ++i){
        // calculate the remaining gas after arriving at station i
        totalGas += (gas[i] - cost[i])
        // If the totalGas < 0, that means it's impossible to arrive at this station - it will be an unsuccessful route.
        if (totalGas < 0){
            start = i + 1
            totalGas = 0
        }
    }
    return start
};

var canCompleteCircuit = function(gas, cost) {
    /*
    This isn't a very intuitive solution, so it's better to stick with the greedy approach above
    Top Down Recursive Solution
    O(N^2) Time 
    O(N) Space
    https://leetcode.com/problems/gas-station/discuss/3014300/DP-%2B-Recursion

    Approach:
    1) initially, loop through all the stations and start our recursion starting at station i

    2) At station i, IF our current tank + gas[cur] >= cost[cur], this means we have enough gas
    to leave the station. Therefore, we want to try the next station i + 1

    The problem states that at each i, we must go in a clockwise direction, meaning that 
    if we're at i, we are limited to i+1,... length-1, and cannot go backwards (i.e i-1)
    Therefore, if we're at i=length-1, in order for the indices to "loop" back around, 
    we need to use the % operator. 
    
    So at each recursive call, we pass in (i+1) % length, and then for the gasAmt,
    we take the gasAmt + gas[cur] - cost[cur] to show the amount of gas we have left after leaving the station

    3) Base Case: 
    If we have enough gas to leave the current station, (gas[cur] + gasAmt >= cost[cur]), 
        If the next station is the starting point((cur+1)%gas.length === startingPoint), 
        this means we were able to complete the cycle. return true

    Memoization:
    This isn't very intuitive, but
    at each cur, we want to store the amount of gas left
    memo[cur] = gasAmt

    if we've already seen this station, AND the amount of gas 
    that we had at the first time we've been to this station exceeds the amount of gas we have now,
    that means that there's no point in continuing, as we're not going to have enough gas to continue
    anyway.

    This doesn't seem like something you could come up with in an interview setting since it 
    does not match how memoization is normally used for recursive/DP style problems, so the greedy approach
    is better.

    */
    let N = gas.length
    let memo = {}
    var search = function(startingPoint, cur, gasAmt){
        if (cur in memo && memo[cur] >= gasAmt){
            return false
        }
        memo[cur] = gasAmt
        if (gas[cur] + gasAmt >= cost[cur]){
            if ((cur + 1) % gas.length === startingPoint){
                return true
            }
            return search(startingPoint, (cur + 1) % gas.length, gasAmt + gas[cur] - cost[cur])
        }
        return false
    }
    for (let i = 0; i < gas.length; ++i){
        if (search(i, i, 0)){
            return i
        }
    }
    return -1

};